src/librustc/session/mod.rs

```diff
@@ -694,27 +697,24 @@ impl Session {
         )
     }
 
-    pub fn sysroot<'a>(&'a self) -> &'a Path {
-        match self.opts.maybe_sysroot {
-            Some(ref sysroot) => sysroot,
-            None => self.default_sysroot
-                        .as_ref()
-                        .expect("missing sysroot and default_sysroot in Session"),
-        }
-    }
     pub fn target_filesearch(&self, kind: PathKind) -> filesearch::FileSearch<'_> {
         filesearch::FileSearch::new(
-            self.sysroot(),
+            &self.sysroot,
             self.opts.target_triple.triple(),
             &self.opts.search_paths,
+            match &self.target_tlib_path {
+                Some(target_tlib_path) => &target_tlib_path,
+                None => &self.host_tlib_path,  // host and target are the same
+            },
```

Comment:
This could be `self.target_tlib_path.as_ref().unwrap_or(&self.host_tlib_path)` if that seems nicer.